File: 0130-surrounded-regions/0130-surrounded-regions.cpp

class Solution {
public:
    int row[4] = {-1, 1, 0, 0};
    int col[4] = {0, 0, -1, 1};

    bool valid(int i, int j, int r, int c) {
        return i >= 0 && j >= 0 && i < r && j < c;
    }

    void solve(vector<vector<char>>& board) {
        if (board.empty()) return;

        int r = board.size();
        int c = board[0].size();
        queue<pair<int,int>> q;

        // First column
        for(int i = 0; i < r; i++){
            if(board[i][0] == 'O'){
                board[i][0] = 'T';
                q.push({i,0});
            }
        }

        // Last column
        for(int i = 0; i < r; i++){
            if(board[i][c-1] == 'O'){
                board[i][c-1] = 'T';
                q.push({i,c-1});
            }
        }

        // First row
        for(int j = 0; j < c; j++){
            if(board[0][j] == 'O'){
                board[0][j] = 'T';
                q.push({0,j});
            }
        }

        // Last row
        for(int j = 0; j < c; j++){
            if(board[r-1][j] == 'O'){
                board[r-1][j] = 'T';
                q.push({r-1,j});
            }
        }

       
        while(!q.empty()){
            auto [i, j] = q.front();
            q.pop();

            for(int k = 0; k < 4; k++){
                int ni = i + row[k];
                int nj = j + col[k];
                if(valid(ni, nj, r, c) && board[ni][nj] == 'O'){
                    board[ni][nj] = 'T';
                    q.push({ni, nj});
                }
            }
        }

        
        for(int i = 0; i < r; i++){
            for(int j = 0; j < c; j++){
                if(board[i][j] == 'O') board[i][j] = 'X';
                else if(board[i][j] == 'T') board[i][j] = 'O';
            }
        }
    }
};

File: 0130-surrounded-regions/0130-surrounded-regions.py
from collections import deque
from typing import List

class Solution:

    def bfs(self, i, j, mat):

        q = deque()

        q.append((i,j))
        mat[i][j] = "T"

        r = len(mat)
        c = len(mat[0])

        row = [-1,1,0,0]
        col = [0,0,-1,1]

        while q:

            x,y = q.popleft()

            for k in range(4):

                nx = x + row[k]
                ny = y + col[k]

                if (0 <= nx < r and 
                    0 <= ny < c and 
                    mat[nx][ny] == "O"):

                    mat[nx][ny] = "T"
                    q.append((nx,ny))


    def solve(self, mat: List[List[str]]) -> None:

        if not mat:
            return

        r = len(mat)
        c = len(mat[0])


        # first row
        for j in range(c):
            if mat[0][j] == "O":
                self.bfs(0,j,mat)


        # last row
        for j in range(c):
            if mat[r-1][j] == "O":
                self.bfs(r-1,j,mat)


        # left column
        for i in range(r):
            if mat[i][0] == "O":
                self.bfs(i,0,mat)


        # right column
        for i in range(r):
            if mat[i][c-1] == "O":
                self.bfs(i,c-1,mat)


        # convert
        for i in range(r):
            for j in range(c):

                if mat[i][j] == "T":
                    mat[i][j] = "O"

                elif mat[i][j] == "O":
                    mat[i][j] = "X"